### ISSUE TITLE
RegisterClassW() multi-threaded issue with static link

### ISSUE THREAD

**Issue by TronicLabs** [glfw/glfw#469]
error:0x10008 description:Win32: Failed to register window class

[win32_window.c#L773](https://github.com/glfw/glfw/blob/master/src/win32_window.c#L773)

this in my case resolve the failure.
but it is really thread safe?

``` c++
    if (!GetClassInfoW(GetModuleHandleW(NULL), _GLFW_WNDCLASSNAME, &wc))
    {
        if (!RegisterClassW(&wc))
        {
            _glfwInputError(GLFW_PLATFORM_ERROR,
                "Win32: Failed to register window class");
            return GL_FALSE;
        }
    }
```

or you can replace all
[win32_window.c#L694](https://github.com/glfw/glfw/blob/master/src/win32_window.c#L694) [win32_window.c#L759](https://github.com/glfw/glfw/blob/master/src/win32_window.c#L759) [win32_window.c#L766](https://github.com/glfw/glfw/blob/master/src/win32_window.c#L766) [win32_window.c#L787](https://github.com/glfw/glfw/blob/master/src/win32_window.c#L787) 

``` c++
GetModuleHandleW(NULL)
```

 with:

``` c++
EXTERN_C IMAGE_DOS_HEADER __ImageBase;
#define HINST_THISCOMPONENT ((HINSTANCE)&__ImageBase)
 ...
wc.hInstance = HINST_THISCOMPONENT; // replace > GetModuleHandleW(NULL)
...
```

this ensure to use the right module hInstances, so we can have same named class, without clash.

and it would also be preferable to give the opportunity to the user to name the windows class.

**Comment by elmindreda** [comment]
That code will only ever be called from the main thread.

**Comment by TronicLabs** [comment]
In my case I have embedded GLFW in some DLL that statically link it,
then the main application load them to create the window,
so the windows class with GetModuleHandleW (NULL) will point to the main exe application,
so aftert loaded the first DLL and create the window,
not allow you to create other window from the other DLL call, becose the class is already registered.
So the my code proposal, resolve this scenario.

**Comment by elmindreda** [comment]
Okay. That has nothing to do with threads, though.

Why not put GLFW in its own DLL and link your ones to that? GLFW wasn't written with multiple copies per process in mind.

**Comment by TronicLabs** [comment]
My DLL per se are mini applications already with its window that are loaded as a plugin,
and I use the GLFW to create additional opengl child window to link with the parent.

I hope that this solution can be adopted on other platforms, to have multiple copies per process.

**Comment by elmindreda** [comment]
I don't see how any of that precludes having GLFW as its own DLL.

**Comment by TronicLabs** [comment]
The glfw3.lib is already static linked with my DLL.
Maybe I did not understand what you mean?

**Comment by TronicLabs** [comment]
Just to clarify the scenario, the main application is not managed by me, it loads my DLL as plugins.

**Comment by elmindreda** [comment]
This should be fixed now with 07a5518c3e42f4412a5d9fa8620950d8f5041e28, which works mostly the same way. Thank you for the bug report and for you patience.
